source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc

```diff
@@ -385,164 +357,109 @@ bool Filter::parseTlvs(const std::vector<uint8_t>& tlvs) {
     }
 
     idx += tlv_value_length;
-    ASSERT(idx <= tlvs.size());
+    ASSERT(idx <= len);
   }
   return true;
 }
 
-ReadOrParseState Filter::readExtensions(Network::IoHandle& io_handle) {
-  // Parse and discard the extensions if this is a local command or there's no TLV needs to be saved
-  // to metadata.
-  if (proxy_protocol_header_.value().local_command_ || 0 == config_->numberOfNeededTlvTypes()) {
-    // buf_ is no longer in use so we re-use it to read/discard.
-    return parseExtensions(io_handle, reinterpret_cast<uint8_t*>(buf_), sizeof(buf_), nullptr);
-  }
-
-  // Initialize the buf_tlv_ only when we need to read the TLVs.
-  if (buf_tlv_.empty()) {
-    buf_tlv_.resize(proxy_protocol_header_.value().extensions_length_);
+ReadOrParseState Filter::readExtensions(Network::ListenerFilterBuffer& buffer) {
+  auto raw_slice = buffer.rawSlice();
+  // waiting for more data if there is no enough data for extensions.
+  if (raw_slice.len_ < (proxy_protocol_header_.value().wholeHeaderLength())) {
+    return ReadOrParseState::TryAgainLater;
   }
 
-  // Parse until we have all the TLVs in buf_tlv.
-  const ReadOrParseState parse_extensions_state =
-      parseExtensions(io_handle, buf_tlv_.data(), buf_tlv_.size(), &buf_tlv_off_);
-  if (parse_extensions_state != ReadOrParseState::Done) {
-    return parse_extensions_state;
+  if (proxy_protocol_header_.value().local_command_ || 0 == config_->numberOfNeededTlvTypes()) {
+    // Ignores the extensions if this is a local command or there's no TLV needs to be saved
+    // to metadata. Those will drained from the buffer in the end.
+    return ReadOrParseState::Done;
   }
 
-  if (!parseTlvs(buf_tlv_)) {
+  const uint8_t* buf = static_cast<const uint8_t*>(raw_slice.mem_) +
+                       proxy_protocol_header_.value().headerLengthWithoutExtension();
+  if (!parseTlvs(buf, proxy_protocol_header_.value().extensions_length_)) {
     return ReadOrParseState::Error;
   }
 
   return ReadOrParseState::Done;
 }
 
-ReadOrParseState Filter::readProxyHeader(Network::IoHandle& io_handle) {
-  while (buf_off_ < MAX_PROXY_PROTO_LEN_V2) {
-    const auto result =
-        io_handle.recv(buf_ + buf_off_, MAX_PROXY_PROTO_LEN_V2 - buf_off_, MSG_PEEK);
-
-    if (!result.ok()) {
-      if (result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again) {
-        return ReadOrParseState::TryAgainLater;
-      }
-      ENVOY_LOG(debug, "failed to read proxy protocol (no bytes read)");
+ReadOrParseState Filter::readProxyHeader(Network::ListenerFilterBuffer& buffer) {
+  auto raw_slice = buffer.rawSlice();
+  const char* buf = static_cast<const char*>(raw_slice.mem_);
```

Comment:
`const uint8_t*`?

Reply:
The original code is using `const char*` for buffer. I guess the reason is that proxy v1 is text based. https://github.com/envoyproxy/envoy/blob/main/source/extensions/common/proxy_protocol/proxy_protocol_header.h#L15-L20

So there are some string operations for parse v1 protocol. 
https://github.com/envoyproxy/envoy/pull/17395/files/ec80e8825245aa0ed651ec63c587aa77e3d53c40#diff-a35fc10b1ce97239d02ac1791aac05069eaf253d8efc77cefe93749f901e36abR239

But I could change that if you still think that is the right thing here.